### PR TITLE
Associate document with user who created it

### PR DESCRIPTION
**Pull request by emmabeynon**
For https://trello.com/c/Th0t8cUb/165-associate-a-document-with-the-user-who-created-it

In this PR we associate a document with a creator so we're able to look up who created it.  This allows us to display information about the person who created the document; for now it will just be the creator's name on the document summary page, but in future could include their organisation for example.

screen shot 2018-08-30 at 14 13 36

We also handle instances where a document has no creator by displaying it as "Unknown".

### CODE REVIEW

File: db/migrate/20180830104245_add_creator_id_to_document.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddCreatorIdToDocument < ActiveRecord::Migration[5.2]
+  def change
+    add_column :documents, :creator_id, :integer
```
Comment (kevindew) [inline db/migrate/20180830104245_add_creator_id_to_document.rb]: It might be better to do this through a [`add_reference`](https://apidock.com/rails/v4.0.2/ActiveRecord/ConnectionAdapters/SchemaStatements/add_reference) since we're referencing the user table, and will match the data type (which I believe is a bigint)

File: app/models/document.rb [hunk 2]
```diff
@@ -22,4 +22,8 @@ def document_type_schema
   def newly_created?
     self.created_at == self.updated_at
   end
+
+  def created_by
+    User.find_by(id: self.creator_id)
+  end
```
Comment (kevindew) [inline app/models/document.rb]: It might be better to use a belongs_to to reference this association?

File: spec/factories/document_factory.rb [hunk 3]
```diff
@@ -8,5 +8,6 @@
     base_path { title ? "/prefix/#{title.parameterize}" : nil }
     document_type { build(:document_type_schema, path_prefix: "/prefix").id }
     publication_state { "changes_not_sent_to_draft" }
+    creator_id { User.last.id }
```
Comment (kevindew) [inline spec/factories/document_factory.rb]: I think if we switch this to be a one to one association we can simplify this rather than needing to pull an id out of the db.

File: config/locales/en/documents/show.yml [hunk 4]
```diff
@@ -32,6 +32,7 @@ en:
         status: Status
         updated_at: Last updated
         created_at: Created
+        created_by: Document created by
```
Comment (kevindew) [inline config/locales/en/documents/show.yml]: We could pass in the name as a variable as per https://github.com/alphagov/content-publisher/blob/b62190d6873105e2cc244dc5e502e980aaa724b7/config/locales/en/validations.yml#L5 and then you wouldn't need to concatenate the string after?

File: app/models/document.rb [hunk 5]
```diff
@@ -22,4 +22,8 @@ def document_type_schema
   def newly_created?
     self.created_at == self.updated_at
   end
+
+  def created_by
```
Comment (tijmenb) [inline app/models/document.rb]: Have you considered setting up a `belongs_to :creator` relationship for this? It would prevent us from having to manually define the `created_by`.

File: db/migrate/20180830104245_add_creator_id_to_document.rb (hunk 1 above)
Comment (tijmenb) [inline db/migrate/20180830104245_add_creator_id_to_document.rb]: Can we add a foreign key to the users table here, so that we won't be able to accidentally delete users that have documents attached?

**Comment by emmabeynon** [comment]
@tijmenb @kevindew I was originally going to do a belongs_to association, but not quite knowing how history is going to work in future I thought it might be simpler and clearer to do it this way and hopefully easy to get rid of once we know what history will look like.  I also wanted to avoid creating a separate model for a creator; it's how it's done in Whitehall and it seemed a bit unnecessary.  Happy to chat through further in any case 👍

File: config/locales/en/documents/show.yml [hunk 6]
```diff
@@ -32,6 +32,8 @@ en:
...
+        unknown_data: Unknown
```
Comment (benthorner) [inline config/locales/en/documents/show.yml]: Tiny thing, but could we call this "unknown_creator", just so it's obvious where it's being shown?

**Comment by kevindew** [comment]
@emmabeynon Yeah hopefully we'll crack history in the not too distant future. The person who changes a document shouldn't change as part of history.

I don't think you'll need to create a creator model, just an association between document and user that is named creator?

File: config/locales/en/documents/show.yml (hunk 6 above)
Comment (emmabeynon) [reply config/locales/en/documents/show.yml]: Sure - my thinking on this was it could be reused for other future metadata fields for which we don't have the data.

File: spec/features/show_a_document_spec.rb [hunk 7]
```diff
@@ -20,5 +20,6 @@ def then_i_see_the_document_summary
     expect(page).to have_content(@document.summary)
     expect(page).to have_content(@document.created_at.strftime("%l:%M%P on %d %B %Y"))
     expect(page).to have_content(@document.updated_at.strftime("%l:%M%P on %d %B %Y"))
+    expect(page).to have_content(I18n.t("documents.show.metadata.created_by") + ": " + @document.creator.name)
```
Comment (benthorner) [inline spec/features/show_a_document_spec.rb]: I think we're not bothering with the label here - just the value is fine.

File: spec/features/show_a_document_with_no_creator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  def given_there_is_a_document_with_no_creator
+    @document = create(:document, creator_id: nil)
+  end
+
+  def when_i_visit_the_document_page
+    visit document_path(@document)
+  end
+
+  def the_document_creator_is_shown_as_unknown
+    expect(page).to have_content(I18n.t("documents.show.metadata.unknown_data"))
```
Comment (benthorner) [inline spec/features/show_a_document_with_no_creator_spec.rb]: This one could be a bit flakey, as there may be other 'Unknown's on the page in future. Could we make the copy include something about users, so the test should always pass for the right reason?

File: app/models/user.rb [hunk 9]
```diff
@@ -2,6 +2,6 @@
 
 class User < ApplicationRecord
   include GDS::SSO::User
-
+  has_many :documents, foreign_key: :creator_id, inverse_of: :user, dependent: :restrict_with_exception
```
Comment (benthorner) [inline app/models/user.rb]: I think we should be OK without the inverse_of, as it's only required when we change the user, which isn't very likely.

File: spec/features/show_a_document_spec.rb (hunk 7 above)
Comment (emmabeynon) [reply spec/features/show_a_document_spec.rb]: Ah so the reason I had to include this was that without it the test would pass as the creator's name would already appear in the page in the top right hand corner.  This helps make sure we're testing the right thing.

File: spec/features/show_a_document_with_no_creator_spec.rb (hunk 8 above)
Comment (emmabeynon) [reply spec/features/show_a_document_with_no_creator_spec.rb]: Good point, I could add the `documents.show.metadata.created_by` microcopy as I did in the other test? What do you think?

File: app/models/user.rb (hunk 9 above)
Comment (emmabeynon) [reply app/models/user.rb]: Rubocop gets upset and demands the `:inverse_of` option if I leave it out 🙈

File: spec/features/show_a_document_spec.rb (hunk 7 above)
Comment (benthorner) [reply spec/features/show_a_document_spec.rb]: So I think you can use a 'within' block (which could go around all of those expectations) to scope the have_content matcher to look within the status box.

File: app/models/user.rb (hunk 9 above)
Comment (benthorner) [reply app/models/user.rb]: Ah, OK. In that case, I think the documentation says it the inverse_of for user should be 'creator' rather than 'user'.

File: spec/features/show_a_document_spec.rb (hunk 7 above)
Comment (emmabeynon) [reply spec/features/show_a_document_spec.rb]: Good idea - will implement that 👍

File: app/models/user.rb (hunk 9 above)
Comment (emmabeynon) [reply app/models/user.rb]: Ah yes good spot, thanks.

**Comment by emmabeynon** [comment]
@benthorner Those changes should be sorted now 👍
